### ISSUE TITLE
[BUG] tw_gtasks_sync Quota Exceeded

### ISSUE THREAD

**Issue by mgiugliano** [bergercookie/syncall#106]
### Describe the bug

Launching tw_gtasks_sync for the first time, completed TW tasks are also synchronised and copied to Google Tasks. I think these are >100 items and I guess Google's API complains.

### To Reproduce

I have launched it just the first time. Launching it again returns an error.

### Expected Behavior

The sync should complete successfully and silently.

### 📸 Screenshots

```
tw_gtasks_sync -p work -l work
23:41:18.82 | INFO     	| Loading preferences...
23:41:18.83 | INFO     	| Loading preferences...
23:41:18.83 | INFO     	| Caching this configuration under the name - work__work__...
23:41:18.83 | INFO     	|

Configuration:
===============

  - TW Tags               : ()
  - TW Project            : work
  - Google Tasks          : work
  - Prefer scheduled dates: False



23:41:18.86 | INFO     	| Loading preferences...
Please visit this URL to authorize this application: https://REDACTED-XXXXXXXXXXXXXXXXX
23:41:30.03 | INFO     	| Creating task list work
23:41:30.30 | INFO     	| Created task list, id: REDACTED
23:41:30.30 | INFO     	| Initializing Taskwarrior...
23:41:31.06 | INFO     	| Detecting changes from Gtasks...
23:41:31.06 | INFO     	| Detecting changes from Tw...
23:41:31.28 | INFO     	| [Tw] Inserting item [XXX] at Gtasks...
:
:
:
23:41:53.80 | WARNING  	|

Google Tasks
------------
	* Items created: 96
	* Items updated: 0
	* Items deleted: 0

Taskwarrior
-----------
	* Items created: 0
	* Items updated: 0
	* Items deleted: 0

23:41:53.80 | ERROR    	| Application failed; Below you can find the error message and stack trace.
```

The last few lines of the error log are:

```
                     -> <syncall.google.gtasks_side.GTasksSide object at 0x10db7d0d0>\n           |    |        -> <bound method Resource._add_nested_resources.<locals>.createResourceMethod.<locals>.methodResource of <googleapiclient.discov...\n           |    -> <googleapiclient.discovery.Resource object at 0x10ddcf950>\n           -> <syncall.google.gtasks_side.GTasksSide object at 0x10db7d0d0>\n  File "/usr/local/lib/python3.11/site-packages/googleapiclient/_helpers.py", line 130, in positional_wrapper\n    return wrapped(*args, **kwargs)\n           |        |       -> {}\n           |        -> (<googleapiclient.http.HttpRequest object at 0x10dfbe5d0>,)\n           -> <function HttpRequest.execute at 0x10d8a8400>\n  File "/usr/local/lib/python3.11/site-packages/googleapiclient/http.py", line 938, in execute\n    raise HttpError(resp, content, uri=self.uri)\n          |         |     |            |    -> \'https://tasks.googleapis.com/tasks/v1/lists/REDACTED/tasks?alt=json\'\n          |         |     |            -> <googleapiclient.http.HttpRequest object at 0x10dfbe5d0>\n          |         |     -> b\'{\\n  "error": {\\n    "code": 403,\\n    "message": "Quota Exceeded",\\n    "errors": [\\n      {\\n        "message": "Quota Ex...\n          |         -> {\'content-type\': \'application/json; charset=UTF-8\', \'vary\': \'Origin, X-Origin, Referer\', \'date\': \'Sun, 28 Jan 2024 22:41:53 G...\n          -> <class \'googleapiclient.errors.HttpError\'>\n\ngoogleapiclient.errors.HttpError: <HttpError 403 when requesting https://tasks.googleapis.com/tasks/v1/lists/REDACTED/tasks?alt=json returned "Quota Exceeded". Details: "[{\'message\': \'Quota Exceeded\', \'domain\': \'usageLimits\', \'reason\': \'quotaExceeded\'}]">\n'
Arguments: ()
```

#### OS Version 

Darwin 23.2.0 Darwin Kernel Version 23.2.0: Wed Nov 15 21:54:10 PST 2023; root:xnu-10002.61.3~2/RELEASE_X86_64 x86_64

#### Python version 

Python 3.11.7

#### `--version` of your executable

tw_gtasks_sync, version 1.7.2

### Additional context

First time syncing...

**Comment by mgiugliano** [comment]
BTW: TW filtering does not work.
e.g.

```
tw_gtasks_sync -l "TW work" -p work --days 30
Usage: tw_gtasks_sync [OPTIONS]
Try 'tw_gtasks_sync --help' for help.

Error: No such option: --days
```

and similarly

```
tw_gtasks_sync -l "TW work" -p work -f "+OVERDUE or +TODAY"
Usage: tw_gtasks_sync [OPTIONS]
Try 'tw_gtasks_sync --help' for help.

Error: No such option: -f
```

**Comment by bergercookie** [comment]
> Launching tw_gtasks_sync for the first time, completed TW tasks are also synchronised and copied to Google Tasks. I think these are >100 items and I guess Google's API complains.

I have reproduced the `Quota Exceeded` error message.

```
<HttpError 403 when requesting https://tasks.googleapis.com/tasks/v1/lists/XzN0bWtIelhfRkQ2X0RrYw/tasks?alt=json returned "Quota Exceeded". Details: "[{'message': 'Quota Exceeded', 'domain': 'usageLimits', 'reason': 'quotaExceeded'}]">
12:13:11.53 | ERROR     | Application failed; Below you can find the error message and stack trace. If you think this is a bug, attach this stack trace to create a new issue -> [https://github.com/bergercookie/syncall/issues].
```

Note: Trying to sync ~100 items makes it possible to reproduce.

However, on subsequent runs, this seems to work fine and sync the rest of the tasks. Is that not the case for you ? Does it still fail with the same `Quota Exceeded` error message ?


> BTW: TW filtering does not work.

This referred to the `master` branch state which hadn't made its way to a pypi release. You can now install from pypi the 1.8.0 release and the new options should be there. Also have a look at the discussion of this new release for caveats: https://github.com/bergercookie/syncall/discussions/107

**Comment by bergercookie** [comment]
Note also that I've just pushed `v1.8.1` where we catch any exceptions thrown during new items addition on either side. In this case, if a `Quota Exceeded` error is thrown, we'll just skip this item momentarily and try with the next, but the application won't exit altogether. ON subsequent runs it will try syncing any items that are still not synchronized.

This fixes it at least on my side. Let me know if this doesn't work for you.

**Comment by mgiugliano** [comment]
Thank you immensely for your quick reaction and effective time you put into it. 
It failed, but this time for some reason might be guilty of.

```
tw_gtasks_sync -l "TW work" -p work
13:23:22.55 | INFO     	| Loading preferences...
13:23:22.56 | INFO     	| Loading preferences...
13:23:22.56 | INFO     	| Caching this configuration under the name - TW%20work____work__.yaml...
13:23:22.56 | INFO     	|

Configuration:
===============

  - TW Filter             :
  - TW Tags               : ()
  - TW Project            : work
  - TW Sync All Tasks     : False
  - Google Tasks          : TW work
  - Prefer scheduled dates: False



13:23:22.56 | INFO     	| Loading preferences...
13:23:22.56 | INFO     	| Using already cached credentials...
13:23:23.01 | INFO     	| Creating task list TW work
13:23:23.21 | INFO     	| Created task list, id: REDACTED
13:23:23.21 | INFO     	| Initializing Taskwarrior...
Exception was raised during program execution.

  File "/usr/local/bin/tw_gtasks_sync", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/syncall/scripts/tw_gtasks_sync.py", line 209, in main
    aggregator.sync()
  File "/usr/local/lib/python3.11/site-packages/syncall/aggregator.py", line 174, in sync
    str(item[self._helper_B.id_key]): item for item in self._side_B.get_all_items()
                                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/syncall/taskwarrior/taskwarrior_side.py", line 152, in get_all_items
    self._load_all_items()
  File "/usr/local/lib/python3.11/site-packages/syncall/taskwarrior/taskwarrior_side.py", line 127, in _load_all_items
    tasks = self._tw.load_tasks_and_filter(command="all", filter_=filter_)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/taskw_ng/warrior.py", line 363, in load_tasks_and_filter
    results = dict(
              ^^^^^
  File "/usr/local/lib/python3.11/site-packages/taskw_ng/warrior.py", line 364, in <genexpr>
    (db, self._get_task_objects(f"status:{db}", query_args, "export"))
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/taskw_ng/warrior.py", line 281, in _get_task_objects
    value = [self._get_task_object(j) for j in json]
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/taskw_ng/warrior.py", line 281, in <listcomp>
    value = [self._get_task_object(j) for j in json]
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/taskw_ng/warrior.py", line 286, in _get_task_object
    return Task(obj, udas=self.config.get_udas())
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/taskw_ng/task.py", line 82, in __init__
    processed[k] = self._deserialize(k, v, self._fields)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/taskw_ng/task.py", line 131, in _deserialize
    return converter.deserialize(value)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/taskw_ng/fields/duration.py", line 116, in deserialize
    return parse_iso8601_duration(value)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/taskw_ng/fields/duration.py", line 66, in parse_iso8601_duration
    raise ValueError(


P360D is not an ISO8601 duration, expected to find the "T" character
13:23:23.58 | ERROR    	| Application failed; Below you can find the error message and stack trace. If you think this is a bug, attach this stack trace to create a new issue -> [https://github.com/bergercookie/syncall/issues]. Also consider running in verbose mode '--verbose' to get more details on this issue.
NoneType: None
```

I got the same error when invoking with other options (i.e. the TW filters), so it seems I have a task with strange "duration".

 
BTW I forgot (as I noticed it only now) that yesterday pip complained with the following versions:

```
syncall 1.7.2 requires bidict<0.22.0,>=0.21.2, but you have bidict 0.22.1 which is incompatible.
syncall 1.7.2 requires loguru<0.6.0,>=0.5.3, but you have loguru 0.7.2 which is incompatible.
syncall 1.7.2 requires PyYAML<5.4.0,>=5.3.1, but you have pyyaml 6.0.1 which is incompatible.
```

As I uninstalled and pip install it again (getting the 1.8.1 version), those warning where gone.

**Comment by bergercookie** [comment]
> P360D is not an ISO8601 duration, expected to find the "T" character

This seems like a bug with `taskw_ng`. P360D is a valid ISO8601 duration so I think we should support it :)

**Comment by bergercookie** [comment]
> 
Thank you immensely for your quick reaction and effective time you put into it.
It failed, but this time for some reason might be guilty of.

tw_gtasks_sync -l "TW work" -p work
13:23:22.55 | INFO     	| Loading preferences...
13:23:22.56 | INFO     	| Loading preferences...
13:23:22.56 | INFO     	| Caching this configuration under the name - TW%20work____work__.yaml...
13:23:22.56 | INFO     	|

Configuration:
===============

  - TW Filter             :
  - TW Tags               : ()
  - TW Project            : work
  - TW Sync All Tasks     : False
  - Google Tasks          : TW work
  - Prefer scheduled dates: False



13:23:22.56 | INFO     	| Loading preferences...
13:23:22.56 | INFO     	| Using already cached credentials...
13:23:23.01 | INFO     	| Creating task list TW work
13:23:23.21 | INFO     	| Created task list, id: REDACTED
13:23:23.21 | INFO     	| Initializing Taskwarrior...
Exception was raised during program execution.

  File "/usr/local/bin/tw_gtasks_sync", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/syncall/scripts/tw_gtasks_sync.py", line 209, in main
    aggregator.sync()
  File "/usr/local/lib/python3.11/site-packages/syncall/aggregator.py", line 174, in sync
    str(item[self._helper_B.id_key]): item for item in self._side_B.get_all_items()
                                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/syncall/taskwarrior/taskwarrior_side.py", line 152, in get_all_items
    self._load_all_items()
  File "/usr/local/lib/python3.11/site-packages/syncall/taskwarrior/taskwarrior_side.py", line 127, in _load_all_items
    tasks = self._tw.load_tasks_and_filter(command="all", filter_=filter_)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/taskw_ng/warrior.py", line 363, in load_tasks_and_filter
    results = dict(
              ^^^^^
  File "/usr/local/lib/python3.11/site-packages/taskw_ng/warrior.py", line 364, in <genexpr>
    (db, self._get_task_objects(f"status:{db}", query_args, "export"))
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/taskw_ng/warrior.py", line 281, in _get_task_objects
    value = [self._get_task_object(j) for j in json]
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/taskw_ng/warrior.py", line 281, in <listcomp>
    value = [self._get_task_object(j) for j in json]
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/taskw_ng/warrior.py", line 286, in _get_task_object
    return Task(obj, udas=self.config.get_udas())
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/taskw_ng/task.py", line 82, in __init__
    processed[k] = self._deserialize(k, v, self._fields)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/taskw_ng/task.py", line 131, in _deserialize
    return converter.deserialize(value)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/taskw_ng/fields/duration.py", line 116, in deserialize
    return parse_iso8601_duration(value)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/taskw_ng/fields/duration.py", line 66, in parse_iso8601_duration
    raise ValueError(


P360D is not an ISO8601 duration, expected to find the "T" character
13:23:23.58 | ERROR    	| Application failed; Below you can find the error message and stack trace. If you think this is a bug, attach this stack trace to create a new issue -> [https://github.com/bergercookie/syncall/issues]. Also consider running in verbose mode '--verbose' to get more details on this issue.
NoneType: None
I got the same error when invoking with other options (i.e. the TW filters), so it seems I have a task with strange "duration".

BTW I forgot (as I noticed it only now) that yesterday pip complained with the following versions:

syncall 1.7.2 requires bidict<0.22.0,>=0.21.2, but you have bidict 0.22.1 which is incompatible.
syncall 1.7.2 requires loguru<0.6.0,>=0.5.3, but you have loguru 0.7.2 which is incompatible.
syncall 1.7.2 requires PyYAML<5.4.0,>=5.3.1, but you have pyyaml 6.0.1 which is incompatible.
As I uninstalled and pip install it again (getting the 1.8.1 version), those warning where gone.



:+1:

**Comment by bergercookie** [comment]
> This seems like a bug with taskw_ng. P360D is a valid ISO8601 duration so I think we should support it :)

This should now be working correctly in syncall 1.8.4 which uses latest taskw-ng 0.2.6.
Do let me know if it doesn't though and I'll have a look :)

**Comment by mgiugliano** [comment]
It works! However it does NOT sync the date/time of TW tasks. Therefore any option related to (e.g.) days appears to fail.

```
tw_gtasks_sync -l "TW work" -p work --days 10
```

This resulted in syncing ALL the "work" project from TW, with >100 tasks completed years ago. When looking at the attributes of the synced tasks from within Google Task, the date/time is completely missing.

I can open a separate issue, if needed. Of course I failed to note whether this was also a bug yesterday, before the improvements you made.

Thanks again for your work and for sync all. I look forward to use it daily (as soon as I understand these glitches).

**Comment by bergercookie** [comment]
> tw_gtasks_sync -l "TW work" -p work --days 10
> This resulted in syncing ALL the "work" project from TW, with >100 tasks completed years ago. 

This is probably because the tasks + days filter are OR-ed. This is the resulting filter using `-p work --days 10`: `+work or modified.after:-1d`. You can see the resulting filter if you run in verbose mode, with `-v` (`-vv`, -vvv`, etc.)

If you want to sync all the `+work` tasks only modified in last 10 days, you can probably use a custom filter like this:

```
-f and modified.after:-10d
```

**Comment by bergercookie** [comment]
Hi @mgiugliano , I've just pushed a fix where I'm basically AND-ing the filters instead of OR-ing them. This should make the `-p work --days 10` combination work since it will filter items that have the `work` tag and are modified in the last 10 days.

The fix should make its way to the 1.8.5 PyPI version momentarily
Let me know if that doesn't work

**Comment by mgiugliano** [comment]
Thank you @bergercookie ! I think the "AND" logic is way more intuitive than the "OR". I apologise for my own confusion and not realising immediately the Google Tasks limitations.
